export/export.py

```diff
@@ -0,0 +1,866 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Union
+
+import torch
+from executorch.devtools.backend_debug import get_delegation_info
+from executorch.exir._warnings import experimental
+from executorch.exir.program import (
+    EdgeProgramManager,
+    ExecutorchProgramManager,
+    to_edge_transform_and_lower,
+)
+from executorch.exir.schema import Program
+from executorch.runtime import Runtime, Verification
+from tabulate import tabulate
+from torch import nn
+from torch.ao.quantization import allow_exported_model_train_eval
+from torch.ao.quantization.quantize_pt2e import convert_pt2e, prepare_pt2e
+from torch.export import export_for_training, ExportedProgram
+from torchao.quantization import quantize_
+from torchao.utils import unwrap_tensor_subclass
+
+from .recipe import ExportRecipe
+
+
+class Stage(ABC):
+    """
+    Interface for a Stage in the ExecuTorch export pipeline.
+
+    Each stage can be connected to other stages to form a pipeline.
+    Stages have clear run and get_outputs functions to make the data flow explicit.
+    Each stage implements its own run method with specific parameter names.
+    """
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """
+        Returns the name of this stage.
+        """
+        pass
+
+    @abstractmethod
+    def run(self, **kwargs) -> None:
+        """
+        Executes this stage with the given inputs.
+
+        Each concrete stage class implements this method with specific parameter names.
+        """
+        pass
+
+    @abstractmethod
+    def get_outputs(self) -> Any:
+        """
+        Returns the outputs generated by this stage.
+
+        Returns:
+            The outputs of this stage, to be used as inputs for the next stage
+        """
+        pass
+
+
+class ExportStage(Stage):
+    """
+    First stage: Export PyTorch model to ExportedProgram.
+    """
+
+    def __init__(
+        self,
+        pre_edge_transform_passes: Optional[
+            Callable[[ExportedProgram], ExportedProgram]
+        ] = None,
+    ) -> None:
+        self._exported_program: Dict[str, ExportedProgram] = {}
+        self._pre_edge_transform_passes = pre_edge_transform_passes
+        self._model_dict: Dict[str, nn.Module] = {}
```

Comment:
is this to capture more than one Module to be exported? why do we need this?

Reply:
Yep a user could possibly pass in multiple methods to be exported and wanted to have support for that from the beginning. Something we struggled with in ModAI by trying to add it later.